cdap-app-templates/cdap-etl/cdap-etl-tools/src/main/java/co/cask/cdap/etl/tool/UpgradeTool.java

```diff
@@ -104,20 +120,144 @@ private boolean upgrade(Id.Application appId) throws Exception {
       return false;
     }
 
+    System.out.println("Upgrading pipeline: " + appId);
     if (BATCH_NAME.equals(appDetail.getArtifact().getName())) {
       OldETLBatchConfig oldConfig = GSON.fromJson(appDetail.getConfiguration(), OldETLBatchConfig.class);
-      ETLBatchConfig newConfig = oldConfig.getNewConfig();
+      ETLBatchConfig newConfig = oldConfig.getNewConfig(batchPluginArtifactFinder);
       AppRequest<ETLBatchConfig> updateRequest = new AppRequest<>(batchArtifact, newConfig);
       appClient.update(appId, updateRequest);
     } else {
       OldETLRealtimeConfig oldConfig = GSON.fromJson(appDetail.getConfiguration(), OldETLRealtimeConfig.class);
-      ETLRealtimeConfig newConfig = oldConfig.getNewConfig();
+      ETLRealtimeConfig newConfig = oldConfig.getNewConfig(realtimePluginArtifactFinder);
       AppRequest<ETLRealtimeConfig> updateRequest = new AppRequest<>(realtimeArtifact, newConfig);
       appClient.update(appId, updateRequest);
     }
     return true;
   }
 
+  // upgrade data the UI stores. None of this should exist ...
+  // UI uses undocumented api in ConsoleSettingsHttpHandler
+  private void upgradeUIData() throws Exception {
+    // UI uses the undocumented console store. this thing stores an arbitrary json object. And its one json object.
+    // UI stores drafts in 'adapterDrafts'
+    // UI stores plugin templates in 'pluginTemplates'
+    // don't know what else is stored in here, so decode the entire blob as a JSONObject to preserve the structure.
+    // then decode adapterDrafts and pluginTemplates separately
+
+    boolean upgradedSomething = false;
+    JsonObject consoleData = consoleClient.get().getAsJsonObject("property");
+    if (consoleData.has("adapterDrafts")) {
+      // sigh... if the 3.3.x UI has started up and been accessed before this tool has run,
+      // the first layer here will be the default namespace.
+      // If the UI has not started up yet, it will be in the 3.2.x format without namespace.
+      //
+      // to recap, in 3.2:
+      //
+      // "adapterDrafts": {
+      //   "TwitterToHBase": { ... }
+      // }
+      //
+      // but if the 3.3 UI has started already:
+      //
+      // "adapterDrafts": {
+      //   "default": {
+      //     "TwitterToHBase": { ... }
+      //   },
+      //   ...,
+      //   "isMigrated": true
+      // }
+      //
+      // Note the 'isMigrated' at the bottom which is a boolean and not an object like all the other keys...
+      Map<String, Map<String, PipelineDraft>> drafts = new HashMap<>();
+      try {
+        // try to deserialize with 3.2 format
+        Map<String, PipelineDraft> defaultDrafts =
+          GSON.fromJson(consoleData.get("adapterDrafts"),
+                        new TypeToken<Map<String, PipelineDraft>>() { }.getType());
+        drafts.put("default", defaultDrafts);
```

Comment:
so there's no way that in 3.2, users could have created drafts in another namespace? I understand though that even if they did, it would be impossible to know which namespace they belonged to. Perhaps add a known issue somewhere, unless I missed something.